group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1087,1348 @@ public void replay(
+
+                List<Record> records = Collections.singletonList(
+                    RecordHelpers.newGroupMetadataRecord(group, metadataImage.features().metadataVersion())
+                );
+
+                return new CoordinatorResult<>(records, appendFuture);
+
+            } else {
+                return maybePrepareRebalance(
+                    group,
+                    "Group's selectedProtocol will change because static member " +
+                        member.memberId() + " with instance id " + groupInstanceId +
+                        " joined with change of protocol; " + "client reason: " + joinReason
+                );
+            }
+
+        } else if (group.isInState(COMPLETING_REBALANCE)) {
+            // if the group is in after-sync stage, upon getting a new join-group of a known static member
+            // we should still trigger a new rebalance, since the old member may already be sent to the leader
+            // for assignment, and hence when the assignment gets back there would be a mismatch of the old member id
+            // with the new replaced member id. As a result the new member id would not get any assignment.
+            return prepareRebalance(group,
+                "Updating metadata for static member " + member.memberId() + " with instance id " +
+                    request.groupInstanceId() + "; client reason: " + joinReason);
+
+        } else if (group.isInState(EMPTY) || group.isInState(DEAD)) {
+            throw new IllegalStateException("Group " + group.groupId() + " was not supposed to be in the state " +
+                group.stateAsString() + " when the unknown static member " + request.groupInstanceId() + " rejoins.");
+
+        } // do nothing if in PREPARING_REBALANCE state.
+        return EMPTY_RESULT;
+    }
+
+    /**
+     * Generate a heartbeat key for the timer.
+     *
+     * Package private for testing.
+     *
+     * @param groupId   The group id.
+     * @param memberId  The member id.
+     *
+     * @return the heartbeat key.
+     */
+    static String heartbeatKey(String groupId, String memberId) {
+        return "heartbeat-" + groupId + "-" + memberId;
+    }
+
+    /**
+     * Generate a join key for the timer.
+     *
+     * Package private for testing.
+     *
+     * @param groupId   The group id.
+     *
+     * @return the join key.
+     */
+    static String joinKey(String groupId) {
+        return "join-" + groupId;
+    }
+
+    /**
+     * Generate a sync key for the timer.
+     *
+     * Package private for testing.
+     *
+     * @param groupId   The group id.
+     *
+     * @return the sync key.
+     */
+    static String syncKey(String groupId) {
+        return "sync-" + groupId;
+    }
+
+    /**
+     * Used for testing.
+     *
+     * @return the new member join timeout in milliseconds.
+     */
+    public int genericGroupNewMemberJoinTimeoutMs() {
+        return this.genericGroupNewMemberJoinTimeoutMs;
+    }
```

Comment:
Could we remove this? `genericGroupNewMemberJoinTimeoutMs` is very likely passed to this object by the test itself so I am not sure to understand why we need to expose it to the test. Is there a reason?